src/com/amazon/corretto/crypto/provider/HpkeCipher.java

```diff
@@ -0,0 +1,309 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.corretto.crypto.provider;
+
+import java.security.AlgorithmParameters;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.InvalidParameterSpecException;
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.CipherSpi;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.ShortBufferException;
+
+public class HpkeCipher extends CipherSpi {
+
+  static {
+    Loader.load();
+  }
+
+  private final AmazonCorrettoCryptoProvider provider_;
+
+  private int javaCipherMode_ = 0;
+  private EvpHpkeKey key_;
+  private HpkeParameterSpec params_;
+
+  private final Object lock_ = new Object();
+
+  HpkeCipher(AmazonCorrettoCryptoProvider provider) {
+    Loader.checkNativeLibraryAvailability();
+    provider_ = provider;
+  }
+
+  // Core Native Methods
+  // -------------------
+
+  /**
+   * Performs single-shot HPKE encryption or decryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return number of bytes written to output, and -1 if failed.
+   */
+  private static native int hpkeCipher(
+      long keyHandle,
+      int javaCipherMode,
+      int kemId,
+      int KdfId,
+      int aeadId,
+      byte[] input,
+      int inputOffset,
+      int inputLen,
+      byte[] output,
+      int outputOffset);
+
+  /**
+   * Computes the number of bytes the output buffer needs to be for wrapping and unwrapping.
+   *
+   * <p>For wrapping, the size is greater than the input buffer since the output buffer also needs
+   * to include an AEAD tag and KEM encapsulate.
+   *
+   * <p>For unwrapping, the size is smaller, since the output does not need the AEAD tag or the KEM
+   * encapsulate.
+   */
+  private static native int hpkeOutputSize(
+      int javaCipherMode, int kemId, int KdfId, int aeadId, int inputLen);
+
+  // Core Java Methods
+  // -----------------
+
+  @Override
+  protected void engineInit(int opmode, Key key, AlgorithmParameterSpec params, SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    if ((opmode != Cipher.WRAP_MODE) && (opmode != Cipher.UNWRAP_MODE)) {
+      throw new IllegalStateException("HpkeCipher only supports WRAP_MODE and UNWRAP_MODE");
+    }
```

Comment:
This limitation is not made clear anywhere other than the source code. Should be apparent in the PR, Javadoc, and README.

Reply:
I ended up removing this limitation.